In [1]:
import pandas as pd
import sqlite3

pd.set_option("display.max_columns", None)

In [2]:
birds_df = pd.read_csv(
    "../cleaned_data/birds_cleaned.csv"
)

birds_df.head()

,AOU_Code,AcceptedTSN,Admin_Sheet,Admin_Unit_Code,Common_Name,Date,Distance,Disturbance,End_Time,Flyover_Observed,Habitat,Humidity,ID_Method,Initial_Three_Min_Cnt,Interval_Length,Location_Type,NPSTaxonCode,Observer,PIF_Watchlist_Status,Plot_Name,Regional_Stewardship_Status,Scientific_Name,Sex,Site_Name,Sky,Start_Time,Temperature,Visit,Wind,Year,Month,Month_Name,Quarter,Season
0,EATO,179276.0,ANTI,ANTI,Eastern Towhee,2018-05-22,<= 50 Meters,No effect on count,06:29:00,False,Forest,79.400002,Singing,True,0-2.5 min,Forest,83803.0,Elizabeth Oswald,False,ANTI-0036,True,Pipilo erythrophthalmus,Undetermined,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring
1,WBNU,178775.0,ANTI,ANTI,White-breasted Nuthatch,2018-05-22,<= 50 Meters,No effect on count,06:29:00,False,Forest,79.400002,Calling,True,0-2.5 min,Forest,90935.0,Elizabeth Oswald,False,ANTI-0036,False,Sitta carolinensis,Unknown,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring
2,RBWO,178195.0,ANTI,ANTI,Red-bellied Woodpecker,2018-05-22,50 - 100 Meters,No effect on count,06:29:00,False,Forest,79.400002,Calling,False,2.5 - 5 min,Forest,84865.0,Elizabeth Oswald,False,ANTI-0036,False,Melanerpes carolinus,Unknown,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring
3,OROR,179064.0,ANTI,ANTI,Orchard Oriole,2018-05-22,<= 50 Meters,No effect on count,06:29:00,False,Forest,79.400002,Singing,False,2.5 - 5 min,Forest,93634.0,Elizabeth Oswald,False,ANTI-0036,False,Icterus spurius,Unknown,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring
4,NOMO,178620.0,ANTI,ANTI,Northern Mockingbird,2018-05-22,<= 50 Meters,No effect on count,06:29:00,False,Forest,79.400002,Visualization,False,2.5 - 5 min,Forest,88394.0,Elizabeth Oswald,False,ANTI-0036,False,Mimus polyglottos,Unknown,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring


In [3]:
conn = sqlite3.connect(
    "../database/bird_species.db"
)

print("Database Connected")

Database Connected


In [4]:
birds_df.to_sql(
    "bird_observations",
    conn,
    if_exists="replace",
    index=False
)

print("Table Created Successfully")

Table Created Successfully


In [5]:
query = """
SELECT *
FROM bird_observations
LIMIT 5
"""

pd.read_sql(query, conn)

,AOU_Code,AcceptedTSN,Admin_Sheet,Admin_Unit_Code,Common_Name,Date,Distance,Disturbance,End_Time,Flyover_Observed,Habitat,Humidity,ID_Method,Initial_Three_Min_Cnt,Interval_Length,Location_Type,NPSTaxonCode,Observer,PIF_Watchlist_Status,Plot_Name,Regional_Stewardship_Status,Scientific_Name,Sex,Site_Name,Sky,Start_Time,Temperature,Visit,Wind,Year,Month,Month_Name,Quarter,Season
0,EATO,179276.0,ANTI,ANTI,Eastern Towhee,2018-05-22,<= 50 Meters,No effect on count,06:29:00,0,Forest,79.400002,Singing,1,0-2.5 min,Forest,83803.0,Elizabeth Oswald,0,ANTI-0036,1,Pipilo erythrophthalmus,Undetermined,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring
1,WBNU,178775.0,ANTI,ANTI,White-breasted Nuthatch,2018-05-22,<= 50 Meters,No effect on count,06:29:00,0,Forest,79.400002,Calling,1,0-2.5 min,Forest,90935.0,Elizabeth Oswald,0,ANTI-0036,0,Sitta carolinensis,Unknown,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring
2,RBWO,178195.0,ANTI,ANTI,Red-bellied Woodpecker,2018-05-22,50 - 100 Meters,No effect on count,06:29:00,0,Forest,79.400002,Calling,0,2.5 - 5 min,Forest,84865.0,Elizabeth Oswald,0,ANTI-0036,0,Melanerpes carolinus,Unknown,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring
3,OROR,179064.0,ANTI,ANTI,Orchard Oriole,2018-05-22,<= 50 Meters,No effect on count,06:29:00,0,Forest,79.400002,Singing,0,2.5 - 5 min,Forest,93634.0,Elizabeth Oswald,0,ANTI-0036,0,Icterus spurius,Unknown,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring
4,NOMO,178620.0,ANTI,ANTI,Northern Mockingbird,2018-05-22,<= 50 Meters,No effect on count,06:29:00,0,Forest,79.400002,Visualization,0,2.5 - 5 min,Forest,88394.0,Elizabeth Oswald,0,ANTI-0036,0,Mimus polyglottos,Unknown,ANTI 1,Cloudy/Overcast,06:19:00,19.9,1,Calm (< 1 mph) smoke rises vertically,2018,5,May,2,Spring


In [6]:
query = """
SELECT COUNT(*) AS Total_Observations
FROM bird_observations
"""

pd.read_sql(query, conn)

,Total_Observations
0,15372


In [7]:
query = """
SELECT COUNT(DISTINCT Scientific_Name)
AS Total_Species
FROM bird_observations
"""

pd.read_sql(query, conn)

,Total_Species
0,127


In [8]:
query = """
SELECT
    Common_Name,
    COUNT(*) AS Observations
FROM bird_observations
GROUP BY Common_Name
ORDER BY Observations DESC
LIMIT 10
"""

pd.read_sql(query, conn)

,Common_Name,Observations
0,Northern Cardinal,1125
1,Carolina Wren,993
2,Red-eyed Vireo,738
3,Eastern Tufted Titmouse,720
4,Indigo Bunting,611
5,Eastern Wood-Pewee,574
6,Field Sparrow,492
7,Red-bellied Woodpecker,489
8,American Robin,470
9,Acadian Flycatcher,462


In [9]:
query = """
SELECT
    Habitat,
    COUNT(*) AS Observations
FROM bird_observations
GROUP BY Habitat
"""

pd.read_sql(query, conn)

,Habitat,Observations
0,Forest,8546
1,Grassland,6826


In [10]:
query = """
SELECT
    Admin_Unit_Code,
    COUNT(*) AS Observations
FROM bird_observations
GROUP BY Admin_Unit_Code
ORDER BY Observations DESC
"""

pd.read_sql(query, conn)

,Admin_Unit_Code,Observations
0,ANTI,3463
1,MONO,2522
2,PRWI,2463
3,CHOH,2202
4,MANA,1901
5,CATO,805
6,NACE,684
7,HAFE,530
8,GWMP,386
9,ROCR,289


In [11]:
query = """
SELECT
    Habitat,
    COUNT(DISTINCT Scientific_Name)
    AS Unique_Species
FROM bird_observations
GROUP BY Habitat
"""

pd.read_sql(query, conn)

,Habitat,Unique_Species
0,Forest,108
1,Grassland,107


In [12]:
query = """
SELECT
    Flyover_Observed,
    COUNT(*) AS Count
FROM bird_observations
GROUP BY Flyover_Observed
"""

pd.read_sql(query, conn)

,Flyover_Observed,Count
0,0,14683
1,1,689


## SQL Analysis Findings

### 1. Dataset Overview
- Total observations: 15,372
- Total unique bird species: 127

### 2. Most Frequently Observed Species
Top observed species:

1. Northern Cardinal (1125)
2. Carolina Wren (993)
3. Red-eyed Vireo (738)
4. Eastern Tufted Titmouse (720)
5. Indigo Bunting (611)

### 3. Habitat Distribution
- Forest: 8,546 observations
- Grassland: 6,826 observations

Forest habitats generated more observations overall.

### 4. Administrative Unit Activity
Top monitoring units:

1. ANTI (3463)
2. MONO (2522)
3. PRWI (2463)
4. CHOH (2202)
5. MANA (1901)

ANTI was the most active observation region.

### 5. Species Diversity
- Forest: 108 species
- Grassland: 107 species

Species richness is nearly identical between habitats.

### 6. Flyover Analysis
- Non-flyover observations: 14,683
- Flyover observations: 689

Most bird detections were associated with habitat occupancy rather than flyover activity.